control-plane/connect-inject/endpoints_controller_test.go

```diff
@@ -5035,6 +5099,99 @@ func TestMapAddresses(t *testing.T) {
+			podName:          "terminating-gateway",
+			serviceName:      "terminating-gateway",
+			serviceNamespace: "consul",
+			annotations: map[string]string{
+				annotationGatewayKind: "terminating",
+			},
+		},
+		"Terminating Gateway with Consul Namespace": {
+			podName:          "terminating-gateway",
+			serviceName:      "terminating-gateway",
+			serviceNamespace: "consul",
+			annotations: map[string]string{
+				annotationGatewayKind:                 "terminating",
+				annotationTerminatingGatewayNamespace: "consul-namespace",
+			},
+		},
+		"Terminating Gateway with Metrics Enabled": {
+			podName:          "terminating-gateway",
+			serviceName:      "terminating-gateway",
+			serviceNamespace: "consul",
+			annotations: map[string]string{
+				annotationGatewayKind: "terminating",
+			},
+			metricsConfig: MetricsConfig{
+				DefaultEnableMetrics: true,
+				EnableGatewayMetrics: true,
+			},
+		},
+	}
+
+	for name, c := range cases {
+		t.Run(name, func(t *testing.T) {
+			pod := corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace:   "consul",
+					Name:        c.podName,
+					Annotations: c.annotations,
+				},
+				Status: corev1.PodStatus{
+					PodIP: "1.2.3.4",
+				},
+			}
+			serviceEndpoint := corev1.Endpoints{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: c.serviceNamespace,
+					Name:      c.serviceName,
+				},
+			}
+
+			r := EndpointsController{
+				MetricsConfig: c.metricsConfig,
+			}
+
+			registration, err := r.createGatewayRegistrations(pod, serviceEndpoint, api.HealthPassing)
+			require.NoError(t, err)
+
+			// TODO These will always be true for terminating-gateway, parameterize them when merging with Ashwin.
+			require.Equal(t, api.ServiceKindTerminatingGateway, registration.Service.Kind)
+			require.Equal(t, "terminating-gateway", registration.Service.Service)
+			require.Equal(t, 8443, registration.Service.Port)
+
+			require.Equal(t, c.podName, registration.Service.ID)
+			require.Equal(t, "1.2.3.4", registration.Service.Address)
+			require.Equal(t, c.podName, registration.Service.Meta[MetaKeyPodName])
+			require.Equal(t, c.serviceName, registration.Service.Meta[MetaKeyKubeServiceName])
+			require.Equal(t, c.serviceNamespace, registration.Service.Meta[MetaKeyKubeNS])
+
+			require.Equal(t, fmt.Sprintf("%s/%s", "consul", c.podName), registration.Check.CheckID)
+			require.Equal(t, api.HealthPassing, registration.Check.Status)
+			require.Equal(t, c.podName, registration.Check.ServiceID)
+
+			if value, ok := c.annotations[annotationConsulNamespace]; ok {
+				require.Equal(t, value, registration.Service.Namespace)
+			}
+
+			if c.metricsConfig.DefaultEnableMetrics && c.metricsConfig.EnableGatewayMetrics {
+				require.Equal(t, "1.2.3.4:20200", registration.Service.Proxy.Config["envoy_prometheus_bind_addr"])
+			}
+		})
+	}
+}
+
 func createPod(name, ip string, inject bool, managedByEndpointsController bool) *corev1.Pod {
 	pod := &corev1.Pod{
 		ObjectMeta: metav1.ObjectMeta{
```

Comment:
Could we also test the deregistration code path and check that the ACL token gets deleted?